SLSA v1 terminology: add Package Model (#664)
slsa-framework/slsa: docs/spec/v1.0/terminology.md

Before diving into the [SLSA Levels](levels.md), we need to establish a core set
of terminology and models to describe what we're protecting.

## TODO: Terms we still need to define

> **TODO:** Define these terms before the v1.0 release.

-   Ecosystem
-   Project

> 

Author: Oh, thanks for spotting this! Sent out #718 to fix.

For examples on how this model applies to real-world build systems, see [index
of build types](/provenance/v1#index-of-build-types).

### Package model

Software is distributed in identifiable units called <dfn>packages</dfn>
according the the rules and conventions of a <dfn>package ecosystem</dfn>.
Examples of formal ecosystems include [Python/PyPA](https://www.pypa.io),
[Debian/Apt](https://wiki.debian.org/DebianRepository/Format), and
[OCI](https://github.com/opencontainers/distribution-spec), while examples of

> informal ecosystems include a linking to a file on a website or distributing

Reviewer: ```suggestion
informal ecosystems include linking to a file on a website or distributing
```
"a linking to ..." reads a bit awkwardly

Reviewer 2: I'd suggest "a link".

artifact, the software producer asks the registry to update this mapping to
resolve to the new artifact. The important bit is that the registry represents
the entity or entities with the power to alter what artifacts are accepted by
consumers for a given package name. For example, if consumers only accept
packages signed by a particular public key, then it is access to that public key
that serves as the registry.

[^label]: This resolution might include a version number, label, or some other
    selector in addition to the package name, but that is not important to SLSA.

| Term | Description |
| ---- | ----------- |
| Package | An identifiable unit of software intended for distribution, ambiguously meaning either an "artifact" or a "package name". Only use this term when the ambiguity is acceptable or desirable. |
| Package artifact | A file or other immutable object that is intended for distribution. |
| Package ecosystem | A set of rules and conventions governing how packages are distributed, including how clients resolve a package name into one or more specific artifacts. |
| Package manager client | Client-side tooling to interact with a package ecosystem. |

> | Package name | <p>A mutable collection of artifacts that all represent different versions of the same software. This is the primary identifier that consumers use to obtain the software. Therefore, it is the thing being protected by SLSA.<p>A package name is specific to an ecosystem + registry, has an owner/maintainer, is more general than a specific hash or version, and has a “correct” source location. A package ecosystem may group package names into some sort of hierarchy, such as the Group ID in Maven, though SLSA does not have a special term for this. |

Reviewer 3: `Therefore, it is the thing being protected by SLSA.`

Seems like one of the most useful points to emphasize. Is there anything to say that could spell this out a bit more? Or is here not the place?

> | Package name | <p>A mutable collection of artifacts that all represent different versions of the same software. This is the primary identifier that consumers use to obtain the software. Therefore, it is the thing being protected by SLSA.<p>A package name is specific to an ecosystem + registry, has an owner/maintainer, is more general than a specific hash or version, and has a “correct” source location. A package ecosystem may group package names into some sort of hierarchy, such as the Group ID in Maven, though SLSA does not have a special term for this. |

Reviewer 2: ```suggestion
| Package name | <p>The primary identifier referring to a mutable collection of artifacts that all represent different versions of the same software. SLSA protects this identifier from [impersonation? squatting? reuse?].
<p>A package name is specific to an ecosystem + registry and has an official owner/maintainer. It is more general than a specific hash or version and is associated with a “correct” source location. A package ecosystem may group package names into some sort of hierarchy, such as the Group ID in Maven, though SLSA does not have a special term for this. |
```

Author: Reworded the first half and removed the second half (see other comment).

| Package registry | An entity responsible for mapping package names to artifacts within a packaging ecosystem. Most ecosystems support multiple registries, usually a single global registry and multiple private registries. |

> | Publish [a package] | Make an artifact available for use by registering it with the package registry. In technical terms, this means associating an artifact to a package name. This does not necessarily mean fully public; an artifact may be published for only a subset of users, such as internal testing or a closed beta. |

Reviewer 3: ```suggestion
| Publish [a package] | Make an artifact available for use by registering it with the package registry. In technical terms, this means associating an artifact to a package name. This does not necessarily mean making the artifact fully public; an artifact may be published for only a subset of users, such as internal testing or a closed beta. |
```

Ambiguous terms to avoid:

-   *Package repository* --- Could mean either package registry or package name,
    depending on the ecosystem. To avoid confusion, we always use "repository"
    exclusively to mean "source repository", where there is no ambiguity.
-   *Package manager* (without "client") --- Could mean either package
    ecosystem, package registry, or client-side tooling.

### Mapping to real-world ecosystems

Most real-world ecosystems fit the package model above but use different terms.
The table below attempts to document how various ecosystems map to the SLSA
Package model. There are likely mistakes and omissions; corrections and
additions are welcome!

<table>
  <tr>
    <th>Package ecosystem
    <th>Package registry
    <th>Package name
    <th>Package artifact
  <tr>
    <td colspan=4><em>Languages</em>
  <tr>
    <td><a href="https://packaging.python.org/en/latest/specifications/binary-distribution-format/#file-name-convention">Python</a>
    <td><a href="https://packaging.python.org/en/latest/glossary/#term-Package-Index">Index</a>

>     <td><a href="https://packaging.python.org/en/latest/glossary/#term-Project">Project</a>

Reviewer 4: ```suggestion
    <td><a href="https://packaging.python.org/en/latest/glossary/#term-Project">Project Name</a>
```
The "project" is really the entire thing. The "project name" is just the mutable name.

    <td><a href="https://packaging.python.org/en/latest/glossary/#term-Distribution-Package">Distribution</a>
  <tr>

>     <td><a href="https://www.npmjs.com/">npm</a> (Javascript)

Reviewer: ```suggestion
    <td><a href="https://www.npmjs.com/">npm</a> (JavaScript)
```

    <td><a href="https://docs.npmjs.com/about-the-public-npm-registry">Registry</a>
    <td><a href="https://docs.npmjs.com/package-name-guidelines">Package Name</a>
    <td><a href="https://docs.npmjs.com/about-packages-and-modules">Package</a>
  <tr>

>     <td><a href="https://go.dev/ref/mod">go</a>

Reviewer: ```suggestion
    <td><a href="https://go.dev/ref/mod">Go</a>
```

    <td><em>(see below)</em>
    <td><a href="https://go.dev/ref/mod#glos-module-path">Module path</a>
    <td><a href="https://go.dev/ref/mod#glos-module">Module</a>
  <tr>
    <td><a href="https://maven.apache.org/glossary">Maven</a> (Java)
    <td>Repository
    <td>Group ID + Artifact ID
    <td>Artifact
  <tr>
    <td><a href="https://doc.rust-lang.org/cargo/appendix/glossary.html">Cargo</a> (Rust)
    <td><a href="https://doc.rust-lang.org/cargo/appendix/glossary.html#registry">Registry</a>
    <td>Crate name

>     <td><a href="https://doc.rust-lang.org/cargo/appendix/glossary.html#crate">Crate</a> (or <a href="https://doc.rust-lang.org/cargo/appendix/glossary.html#artifact">Artifact</a>?)

Reviewer 2: If I understand your definitions correctly, then this entry should be Artifact. Artifact refers specifically to the compiled binary, whereas Crate is ambiguous and may refer to either the compiled binary or its (bundled?) source code.

Reviewer 4: I agree, I think this is `Artifact`.

  <tr>
    <td><a href="https://docs.microsoft.com/en-us/nuget/nuget-org/overview-nuget-org">NuGet</a> (C#)
    <td>Host
    <td>Project
    <td>Package
  <tr>
    <td><a href="http://neilb.org/2015/09/05/cpan-glossary.html">Perl</a>
    <td>Upload server
    <td>Distribution
    <td>Release (or Distribution)
  <tr>
    <td colspan=4><em>Operating systems</em>
  <tr>
    <td><a href="https://wiki.debian.org/Teams/Dpkg">Dpkg </a> (e.g. Debian)
    <td><em>?</em>
    <td>Package name
    <td>Package
  <tr>
    <td><a href="https://rpm.org">RPM</a> (e.g. Red Hat)

>     <td>Repositry

Reviewer: ```suggestion
    <td>Repository
```

    <td>Package name
    <td>Package
  <tr>
    <td><a href="https://wiki.archlinux.org/title/Pacman">Pacman</a> (e.g. Arch)
    <td>Repository
    <td>Package name
    <td>Package
  <tr>
    <td><a href="https://docs.brew.sh/Manpage">Homebrew</a> (e.g. Mac)
    <td>Repository (Tap)
    <td>Package name (Formula)
    <td>Binary package (Bottle)
  <tr>
    <td><a href="https://docs.flatpak.org/en/latest/introduction.html#terminology">Flatpak</a>
    <td>Repository
    <td>Application
    <td>Bundle
  <tr>
    <td><a href="https://nixos.org/guides/how-nix-works.html">nix</a> (e.g. <a href="https://nixos.org/">NixOS</a>)
    <td><em>?</em>
    <td><a href="https://nixos.org/manual/nix/stable/glossary.html#gloss-store-object">Store Object</a>?
    <td>Package or <a href="https://nixos.org/manual/nix/stable/glossary.html#gloss-derivation">Derivation</a>
  <tr>
    <td colspan=4><em>Storage systems</em>
  <tr>
    <td><a href="https://cloud.google.com/storage/docs/key-terms">GCS</a>

>     <td><em>n/a</em>

Reviewer 4: ```suggestion
    <td><em><a href="https://cloud.google.com/storage/docs/buckets">Bucket</a>.</em>
```
Maybe? Seems to fit to me.

Author: I don't think so? I think Google (or "GCS") might be the registry here, since it is the entity with control. A bucket is just a unit of the package name, isn't it?

    <td>Object name
    <td>Object
  <tr>
    <td><a href="https://github.com/opencontainers/distribution-spec/blob/main/spec.md#definitions">OCI</a>/Docker
    <td>Registry
    <td>Repository
    <td>Object
  <tr>
    <td colspan=4><em>Meta</em>
  <tr>
    <td><a href="https://deps.dev/glossary">deps.dev</a>: <a href="https://deps.dev/glossary#system">System</a>
    <td><a href="https://deps.dev/glossary#packaging-authority">Packaging authority</a>
    <td><a href="https://deps.dev/glossary#package">Package</a>
    <td><em>n/a</em>
  <tr>
    <td><a href="https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst">purl</a>: type
    <td>Namespace
    <td>Name
    <td><em>n/a</em>
</table>

Notes:

-   Go: Uses a significantly different model than any other ecosystem, where the
    package name is a git repository URL. Clients can fetch directly from the
    git repository, in which case there is no "package". But clients usually
    instead fetch a zip file from a module mirror that proxies the git
    repository, and consult with a checksum database to detect tampering. It is

>     not yet clear which of these constitute the "registry", if either.

Reviewer 2: How will we clarify which is the registry? Do we lack information about Go, or is the problem that we need to form an opinion on our definitions or model?

Reviewer 4: I think in this case, the registry would be the (one or more) git repositories, because they are the ones that permit/refuse changes to a given package namespace (in this case, the repo itself) based on ownership/access.

Author: After more thought, I argue that it's the module proxy. I updated the text accordingly. PTAL.
